### PR TITLE
Allow non-flyway migration

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/BaseFlywayMigrateDatabaseCommand.java [hunk 1]
```diff
@@ -110,7 +118,14 @@ public void run(CommandLine theCommandLine) throws ParseException {
 			.filter(StringUtils::isNotBlank)
 			.collect(Collectors.toSet());
 
-		FlywayMigrator migrator = new FlywayMigrator(myMigrationTableName);
+		boolean dontUseFlyway = theCommandLine.hasOption("dont-use-flyway");
+
+		BaseMigrator migrator;
+		if (dontUseFlyway) {
+			migrator = new BruteForceMigrator();
+		} else {
```
Comment (fil512) [inline hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/BaseFlywayMigrateDatabaseCommand.java]: Awesome.  I love the way you resolved this.

File: hapi-fhir-jpaserver-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/FlywayMigrator.java [hunk 2]
```diff
@@ -31,67 +31,44 @@
...
+		setUsername(theDataSource.getUsername());
+		setPassword(theDataSource.getPassword());
 
 		String driverClassName = theDataSource.getDriverClassName();
 		if (driverClassName == null) {
 			ourLog.error(this.getClass().getSimpleName() + " constructed without a database driver");
 		} else {
-			myDriverType = DriverTypeEnum.fromDriverClassName(driverClassName);
-			if (myDriverType == null) {
+			DriverTypeEnum driverType = DriverTypeEnum.fromDriverClassName(driverClassName);
```
Comment (fil512) [inline hapi-fhir-jpaserver-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/FlywayMigrator.java]: There are two different `DriverEnum` classes.  It's confusing that they have the same name.  Do we need two?  If so, maybe we should change the name of one of them so it's less confusing.

File: hapi-fhir-jpaserver-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/FlywayMigrator.java [hunk 3]
```diff
@@ -31,67 +31,44 @@
...
 	private List<FlywayMigration> myTasks = new ArrayList<>();
-	private boolean myDryRun;
-	private boolean myNoColumnShrink;
 
 	public FlywayMigrator(String theMigrationTableName, BasicDataSource theDataSource) {
 		this(theMigrationTableName);
-		myConnectionUrl = theDataSource.getUrl();
-		myUsername = theDataSource.getUsername();
-		myPassword = theDataSource.getPassword();
+		setConnectionUrl(theDataSource.getUrl());
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/FlywayMigrator.java]: I bet there is, but I think I'd rather leave that to a future ticket so as to de-risk this change

File: hapi-fhir-jpaserver-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java [hunk 4]
```diff
@@ -352,23 +352,24 @@ protected void init400() { // 20190401 - 20190814
...
+		version.onTable("HFJ_SPIDX_DATE").modifyColumn("20190814.10", "RES_TYPE").nonNullable().failureAllowed().withType(BaseTableColumnTypeTask.ColumnTypeEnum.STRING, 100);
+		version.onTable("HFJ_SPIDX_STRING").modifyColumn("20190814.11", "RES_TYPE").nonNullable().failureAllowed().withType(BaseTableColumnTypeTask.ColumnTypeEnum.STRING, 100);
 		version.onTable("HFJ_SPIDX_STRING").addColumn("20190814.12", "HASH_IDENTITY").nullable().type(BaseTableColumnTypeTask.ColumnTypeEnum.LONG);
 		version.onTable("HFJ_SPIDX_STRING").addIndex("20190814.13", "IDX_SP_STRING_HASH_IDENT").unique(false).withColumns("HASH_IDENTITY");
-		version.onTable("HFJ_SPIDX_COORDS").modifyColumn("20190814.14", "RES_TYPE").nonNullable().withType(BaseTableColumnTypeTask.ColumnTypeEnum.STRING, 100);
-		version.onTable("HFJ_SPIDX_QUANTITY").modifyColumn("20190814.15", "RES_TYPE").nonNullable().withType(BaseTableColumnTypeTask.ColumnTypeEnum.STRING, 100);
+		version.onTable("HFJ_SPIDX_COORDS").modifyColumn("20190814.14", "RES_TYPE").nonNullable().failureAllowed().withType(BaseTableColumnTypeTask.ColumnTypeEnum.STRING, 100);
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java]: I think we'd only want to use it in cases where we're shrinking columns, setting not-nullable, etc.. like, basically only for cleanup tasks, which is what these ones are

So with that in mind, I don't think it's risky to keep them here

File: hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/BaseFlywayMigrateDatabaseCommand.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/BaseFlywayMigrateDatabaseCommand.java]: Woohoo!

**Pull request merged**